src/lib/tabs/tab-header.ts

```diff
@@ -0,0 +1,258 @@
+  private _disableScrollBefore = true;
+
+  /** The index of the active tab. */
+  private _selectedIndex: number = 0;
+  @Input() set selectedIndex(value: number) {
+    this._selectedIndex = value;
+    this._focusIndex = value;
+
+    // Scroll to this index on the next view check in case the tab has not yet been created.
+    this._scrollToSelectedIndex = true;
+  }
+  get selectedIndex(): number {
+    return this._selectedIndex;
+  }
+
+  /** Event emitted when the option is selected. */
+  @Output() selectFocusedIndex = new EventEmitter();
+
+  /** Event emitted when a label is focused. */
+  @Output() indexFocused = new EventEmitter();
+
+  constructor(private _zone: NgZone,
+              private _elementRef: ElementRef,
+              private _renderer: Renderer,
+              @Optional() private _dir: Dir) {}
+
+  /**
+   * Waits one frame for the view to update, then updates the ink bar and scroll.
+   * Note: This must be run outside of the zone or it will create an infinite change detection loop
+   * TODO: internal
+   */
+  ngAfterViewChecked(): void {
+    this._zone.runOutsideAngular(() => {
+      window.requestAnimationFrame(() => {
+        this._updateInkBar();
+        this._updateScrollPosition();
+
+        this._disableScrollBefore = this.scrollDistance == 0;
+        this._disableScrollAfter = this.scrollDistance == this._getMaxScrollDistance();
+      });
+    });
+
+    if (this._scrollToSelectedIndex) {
+      if (this._isScrollingEnabled()) {
+        this._scrollToLabel(this._selectedIndex);
+      }
+      this._scrollToSelectedIndex = false;
+    }
+  }
+
+  /** Tells the ink-bar to align itself to the current label wrapper */
+  private _updateInkBar(): void {
+    const selectedLabelWrapper = this._labelWrappers && this._labelWrappers.length
+        ? this._labelWrappers.toArray()[this.selectedIndex].elementRef.nativeElement
+        : null;
+    this._inkBar.alignToElement(selectedLabelWrapper);
+  }
+
+  handleKeydown(event: KeyboardEvent) {
+    switch (event.keyCode) {
+      case RIGHT_ARROW:
+        this._focusNextTab();
+        break;
+      case LEFT_ARROW:
+        this._focusPreviousTab();
+        break;
+      case ENTER:
+        this.selectFocusedIndex.emit(this.focusIndex);
+        break;
+    }
+  }
+
+  /**
+   * Determines if an index is valid.  If the tabs are not ready yet, we assume that the user is
+   * providing a valid index and return true.
+   */
+  _isValidIndex(index: number): boolean {
+    if (!this._labelWrappers) { return true; }
+
+    const tab = this._labelWrappers.toArray()[index];
+    return tab && !tab.disabled;
+  }
+
+  /** Tracks which element has focus; used for keyboard navigation */
+  get focusIndex(): number {
+    return this._focusIndex;
+  }
+
+  /** When the focus index is set, we must manually send focus to the correct label */
+  set focusIndex(value: number) {
+    if (!this._isValidIndex(value) || this._focusIndex == value) { return; }
```

Comment:
Can you break most or all of this logic outside of the setter directly into a separate method? I generally want to keep setters relatively small.